### ISSUE TITLE
Message Queue to prevent weird behavior

### ISSUE THREAD

**Issue by DaNeubi** [RogueException/Discord.Net#1461]
Hey there,

my bot currently has 2 commands. `/warn <IUser> <string>` and `/unwarn <IUser> <string>`.
While it's no problem if only one person can warn and unwarn members, it got complicated after I gave those rights to a second person.

Scenario:

User **A** will be warned by myself and another moderator.
If this happens at the same time, I don't havve anything in my code to queue these commands and make them execute one after another. They both get executed at the same time. Thus User **A** will get _Warning 3_ for my command execution and again _Warning 3_ for the moderators command execution.

So far I've tried it with SimpleSemaphore but with this approach I only get the following message in my log: `A MessageReceived handler is blocking the gateway task.`

Is there a built in Message queue?

If not, is there a need for in this project?
Because I really need it and would create a Pull request if finished.

**Comment by k-boyle** [comment]
That's because your have your RunMode as `RunMode.Async`. Otherwise all events run one after another

**Comment by ZargorNET** [comment]
Run it in a new Task.Run() and then use a lock. This should fix it 🤔

**Comment by k-boyle** [comment]
... No, don't do that

**Comment by DaNeubi** [comment]
> 
> 
> That's because your have your RunMode as `RunMode.Async`. Otherwise all events run one after another

But synchronus for a discord bot is like....

So your answer let's me think, there is no such thing as a __command queue__ just yet.

I don't think putting every command in such a synchronus - queue is what I or anyone needs. But to be able to put specific commands into such a queue would be great I think.

**Comment by k-boyle** [comment]
no I'm saying don't execute commands in parallel unless you actually need to, which unless your warn command is doing some spaghetti, it shouldnt need to

**Comment by DaNeubi** [comment]
> no I'm saying don't execute commands in parallel unless you actually need to, which unless your warn command is doing some spaghetti, it shouldnt need to

I think I need to execute the `/warn` command in parallel.

I'm setting a group to a user but before that I need to get his current __Warning ...__ group.
While getting these information another call to the same user will cause my problem that only one warning will be added because the second command will calculate the new warning number before the first command updated it.

I thought about setting a temporary group to the user to check if such an action is in progress but a queue would be a better solution I guess in my case.

**Comment by DaNeubi** [comment]
I've come up with a solution.

There is a class called StaticLock
```CS
using System;
using System.Threading.Tasks;

namespace sahnee_bot.Util
{
    public static class StaticLock
    {
        //Variables

        /// <summary>
        /// Command warn
        /// </summary>
        private static bool _lockCommandWarning = false;
        private static readonly Object _lockObjectCommandWarning = new object();
        
        

        #region Getter

        /// <summary>
        /// Tries to lock the warn command
        /// </summary>
        /// <returns></returns>
        private static bool TryLockCommandWarning()
        {
            lock (_lockObjectCommandWarning)
            {
                if (_lockCommandWarning == false)
                {
                    _lockCommandWarning = true;
                    return _lockCommandWarning;
                }
                return false;                
            }
        }

        #endregion

        #region Setter

        /// <summary>
        /// Unlocks the warn command
        /// </summary>
        /// <returns></returns>
        public static bool UnlockCommandWarning()
        {
            lock (_lockObjectCommandWarning)
            {
                if (_lockCommandWarning)
                {
                    _lockCommandWarning = false;
                    return true;
                }
                if (_lockCommandWarning == false)
                {
                    return true;
                }
                return false;
            }
        }
        

        #endregion

        #region LockAquiration

        /// <summary>
        /// Locks the warn command if possible
        /// otherwise it will wait
        /// </summary>
        /// <param name="delay"></param>
        /// <returns></returns>
        public static async Task AquireWarningAsync(int delay = 100)
        {
            while (!TryLockCommandWarning())
            {
                await Task.Delay(delay);
            }
        }

        #endregion
    }
}
```

To use you make it the following in your command:
```CS
        [Command("warn")]
        [RoleHandling(RoleTypes.WarningBotAdmin)]
        [Summary("Warns a specific user in his current guild")]
        public async Task WarnAsync([Summary("The user that will be warned")]IUser user
            , [Summary("The reason why the user had been warned")]string reason)
            {
                try
                {
                    await StaticLock.AquireWarningAsync();
                    await this.ExecuteWarnAsync(user, reason); //Thats where I packed my logic for the command
                }
                finally
                {
                    StaticLock.UnlockCommandWarning();
                }
            }
```

**Comment by k-boyle** [comment]
bruh

**Comment by k-boyle** [comment]
![image](https://user-images.githubusercontent.com/17115277/76692068-19387e80-6649-11ea-9345-d44c7ed1a946.png)
